src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -435,144 +518,124 @@ public void syncOnEachOperation(boolean syncOnEachOperation) {
+     * and updated with any future translog.
      */
-    private static class OperationIteratorImpl implements org.elasticsearch.index.translog.Translog.OperationIterator {
+    class FsView implements View {
 
-        private final TranslogStream translogStream;
-        private final StreamInput input;
-        private final ESLogger logger;
+        boolean closed;
+        final List<FsChannelReader> orderedTranslogs;
 
-        OperationIteratorImpl(ESLogger logger, TranslogStream translogStream, StreamInput input) {
-            this.translogStream = translogStream;
-            this.input = input;
-            this.logger = logger;
+        FsView(List<FsChannelReader> orderedTranslogs) {
+            assert orderedTranslogs.isEmpty() == false;
+            this.orderedTranslogs = orderedTranslogs;
         }
 
-        /**
-         * Returns the next operation in the translog or <code>null</code> if we reached the end of the stream.
-         */
-        public Translog.Operation next() throws IOException {
-            try {
-                if (translogStream instanceof LegacyTranslogStream) {
-                    input.readInt(); // ignored opSize
+        synchronized void onNewTranslog(FsChannelReader oldCurrent, FsChannelReader newCurrent) throws IOException {
+            if (closed) {
+                // we have to close the new references created for as as we will not hold them
+                IOUtils.close(oldCurrent, newCurrent);
+                return;
+            }
+            FsChannelReader reader = orderedTranslogs.remove(orderedTranslogs.size() - 1);
+            reader.close();
+            orderedTranslogs.add(oldCurrent);
+            orderedTranslogs.add(newCurrent);
+        }
+
+        @Override
+        public synchronized long minTrasnlogId() {
+            return orderedTranslogs.get(0).translogId();
+        }
+
+        @Override
+        public synchronized int estimatedTotalOperations() {
+            int ops = 0;
+            for (FsChannelReader translog : orderedTranslogs) {
+                int tops = translog.estimatedTotalOperations();
+                if (tops == FsChannelReader.UNKNOWN) {
+                    return -1;
                 }
-                return translogStream.read(input);
-            } catch (TruncatedTranslogException | EOFException e) {
-                // ignore, not properly written the last op
-                logger.trace("ignoring translog EOF exception, the last operation was not properly written", e);
-                return null;
-            } catch (IOException e) {
-                // ignore, not properly written last op
-                logger.trace("ignoring translog IO exception, the last operation was not properly written", e);
-                return null;
+                ops += tops;
             }
+            return ops;
         }
 
         @Override
+        public synchronized long sizeInBytes() {
```

Comment:
we need this because we iterate on orderedTranslogs. There is no read lock? this is a view..